File: OpenMP_Samples/Problem/Poisson/F/poisson.f
!      -------------------------------
!      Poisson equation with k=1, f= sin(i*j)
!      ------------------------------


       program main
       use omp_lib
       include 'poisson.inc'

       integer DEBUG
       parameter (DEBUG=1)

       double precision EPS
       parameter (EPS=1.0e-4)

       integer MAX_ITER
       parameter (MAX_ITER=10000)

       integer MAX_HEAT
       parameter (MAX_HEAT=100.0)


       double precision,allocatable,dimension(:,:):: U, U_old
       double precision  t0, t1, t2, t_w
       double precision  h, h_pow, dmax
     
       integer i, j
       integer ii, jj
    
       allocate(U(0:M+1,0:M+1))
       allocate(U_old(0:M+1,0:M+1))


       do j=0, M+1
         do i=0, M+1
           U(i, j) = 0.0d0
           U_old(i, j) = 0.0d0
         enddo  
       enddo
   
c      === Set given temperature ----------------------------------
       do j=0, M+1
         U(0, j) = MAX_HEAT
       enddo 
       do i=1, M
         U(i, 0) = MAX_HEAT - dble(i)/dble(M)*MAX_HEAT
         U(i, M+1) = dble(i)/dble(M)*MAX_HEAT
       enddo
       do j=0, M+1
         U(M+1, j) = 0.0d0
       enddo 
c      === End of setting given temperature ------------------------


c      === Start of Solving Poisson equation ----------------------------
       t1 = omp_get_wtime()

       h = 1.0d0 / (dble(M+1))
       h_pow = h*h

c      === Main loop
       do i=1, MAX_ITER

c        --- perform explicit method ( Gauss-Seidel Method )
         call MyPoisson(U, h_pow, M) 
c        --- compute maximum differences
         call CalcErr(U, U_old, M, dmax)

         if (mod(i,100) .eq. 0) then
            print *, " iter=",i, "dmax=",dmax
         endif

         if (dmax .lt. EPS) then
           print *, i, " Iteration is converged in residual ", EPS
           goto 10
         endif

c        --- copy back to U_old
         do jj=1, M
           do ii=1, M
             U_old(ii, jj) = U(ii, jj) 
           enddo  
         enddo        

       enddo 
c      === End of Main loop

       print *, "Iteration is not converged within " 
     &       ,MAX_ITER, " times." 

  10   continue

       t2 = omp_get_wtime()
       t_w =  t2 - t1 
c      === End of Solving Poisson equation ---------------------------

         print *, "M  = ", M
         print *, "MAX_ITER  = ", MAX_ITER

         print *, "time[sec.] = ", t_w

c        === output results  
!         do ii=1, M
!           do jj=1, M
!              print *, ii, jj, U(ii,jj)
!           enddo
!           print *, ""
!         enddo


       deallocate(U)
       deallocate(U_old)

       stop
       end


       subroutine MyPoisson(U_rhs, h_pow, M)
       use omp_lib

       double precision U_rhs(0:M+1, 0:M+1)
       double precision h_pow
       integer M 

       integer i, j  

!      === u_{i,j} = 1/4 (h^2 f_{i,j} + u_{i,j-1} +u_{i-1,j}+u_{i+1,j}+u_{i,j+1})
       do j=1, M
         do i=1, M
           U_rhs(i, j) = 0.25d0 * 
     &      ( h_pow * dsin(dble(i)*dble(j)) +
     &       U_rhs(i,j-1) + U_rhs(i-1,j) + U_rhs(i+1,j) + U_rhs(i,j+1) )
 
         enddo
       enddo

       return
       end


       subroutine CalcErr(U_rhs, U_lhs, M, dmax)
       use omp_lib

       double precision U_rhs(0:M+1, 0:M+1)
       double precision U_lhs(0:M+1, 0:M+1)
       double precision dmax
       integer M 

       integer i,j
       double precision dtemp 
      

       dmax = 0.0
       do j=1, M
         do i=1, M
           dtemp = dabs(U_rhs(i, j) - U_lhs(i, j))
           if (dtemp .gt. dmax ) dmax = dtemp
         enddo
       enddo 
 
       return
       end

File: OpenMP_Samples/Problem/Poisson/C/poisson.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <omp.h>

//!      == number of dimension of discrete 2D mesh
#define  M          100

//!== number of iterations for Gauss-Seidel Method
#define  MAX_ITER   10000

#define  MAX_HEAT   100.0

#define  DEBUG      1

#define  EPS        1.0e-4

/* Please define the matrices in here */
static double  U[M+2][M+2];
static double  U_old[M+2][M+2];


void MyPoisson(double U_rhs[M+2][M+2], double h_pow); 
void CalcErr(double U_rhs[M+2][M+2], double U_lhs[M+2][M+2], double *dmax);

int main(int argc, char* argv[]) {

     double  t0, t1, t2, t_w;
     double  h, h_pow, dmax;

     int     i, j;      
     int     ii, jj;


     for (i=0; i<M+2; i++) {
       for (j=0; j<M+2; j++) {
         U[i][j] = 0.0;
         U_old[i][j] = 0.0;
       }
     }

     //=== Set given temperature ----------------------------------
     for (j=0; j<M+2; j++) {
       U[0][j] = MAX_HEAT;
     }
     for (i=1; i<M+1; i++) {
       U[i][0] = MAX_HEAT - (double)i/(double)M * MAX_HEAT;
       U[i][M+1] = (double)i/(double)M * MAX_HEAT;
     }
     for (j=0; j<M+2; j++) {
       U[M+1][j] = 0.0;
     }
     //=== End of setting given temperature ------------------------


     //=== Start of Solving Poisson equation ----------------------------
     t1 = omp_get_wtime();

     h = 1.0 / (double)(M+1);
     h_pow = h*h;

     //=== Main loop
     for (i=1; i<=MAX_ITER; i++) {


       // --- perform explicit method ( Gauss-Seidel Method )
       MyPoisson(U, h_pow);
       //--- compute maximum differences
       CalcErr(U, U_old, &dmax);

       if (i%100 == 0) {
         printf("iter= %d dmax= %e  \n", i, dmax);
       }

       if (dmax < EPS) {
         printf(" %d Iteration is converged in residual %e \n", i, EPS);
         goto outloop;
       }

       //--- copy back to U_old
       for (ii=1; ii<M+1; ii++) {
         for (jj=1; jj<M+1; jj++) {
	   U_old[ii][jj] = U[ii][jj];
         }
       }

     } //=== End of Main loop

     printf("Iteration is not converged within %d times. \n",MAX_ITER);


 outloop: 

     t2 = omp_get_wtime();
     t_w =  t2 - t1; 
     //=== End of Solving Poisson equation ----------------------------

       printf("M  = %d \n",M);
       printf("MAX_ITER  = %d \n",MAX_ITER);

       printf("time = %lf [sec.] \n",t_w);


       //=== output results
       /*
       for (ii=1; ii<M+1; ii++) {
         for (jj=1; jj<M+1; jj++) {
	   printf("%d %d %lf \n",ii,jj, U[ii][jj]);
         }
         printf("\n");
       }
       */


     return 0;
}


void MyPoisson(double U_rhs[M+2][M+2], double h_pow)
{

  int     i, j;

  //=== u_{i,j} = 1/4 (h^2 f_{i,j} + u_{i,j-1} +u_{i-1,j}+u_{i+1,j}+u_{i,j+1})
  for (i=1; i<M+1; i++) {
    for (j=1; j<M+1; j++) {
      U_rhs[i][j] = 0.25 * 
        ( h_pow * sin((double)i * (double)j) + 
	  U_rhs[i][j-1] + U_rhs[i-1][j] + U_rhs[i+1][j] + U_rhs[i][j+1] );
      }
    }

}

void CalcErr(double U_rhs[M+2][M+2], double U_lhs[M+2][M+2], double *dmax)
{ 

   int  i,j;
   double dtemp;

   *dmax = 0.0;
   for (i=1; i<M+1; i++) {
     for (j=1; j<M+1; j++) {
       dtemp = fabs(U_rhs[i][j] - U_lhs[i][j]);
       if (dtemp > *dmax ) *dmax = dtemp;
     }
   }

}

